system/Helpers/form_helper.php

```diff
@@ -0,0 +1,954 @@
+        
+        return '<input '._parse_form_attributes($data, $defaults)._attributes_to_string($extra)." />\n";
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_button'))
+{
+    /**
+     * Form Button
+     *
+     * @param	mixed
+     * @param	string
+     * @param	mixed
+     * @return	string
+     */
+    function form_button($data = '', string $content = '', $extra = ''): string
+    {
+        $defaults = [
+                'name' => is_array($data) ? '' : $data,
+                'type' => 'button'
+        ];
+        
+        if (is_array($data) && isset($data['content']))
+        {
+            $content = $data['content'];
+            unset($data['content']); // content is not an attribute
+        }
+        
+        return '<button '._parse_form_attributes($data, $defaults)._attributes_to_string($extra).'>'
+                .$content
+                ."</button>\n";
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_label'))
+{
+    /**
+     * Form Label Tag
+     *
+     * @param	string	The text to appear onscreen
+     * @param	string	The id the label applies to
+     * @param	array	Additional attributes
+     * @return	string
+     */
+    function form_label(string $label_text = '', string $id = '', array $attributes = []): string
+    {
+        $label = '<label';
+        
+        if ($id !== '')
+        {
+            $label .= ' for="'.$id.'"';
+        }
+        
+        if (is_array($attributes) && count($attributes) > 0)
+        {
+            foreach ($attributes as $key => $val)
+            {
+                $label .= ' '.$key.'="'.$val.'"';
+            }
+        }
+        
+        return $label.'>'.$label_text.'</label>';
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_datalist'))
+{
+    /**
+     * Datalist
+     * 
+     * The <datalist> element specifies a list of pre-defined options for an <input> element.
+     * Users will see a drop-down list of pre-defined options as they input data.
+     * The list attribute of the <input> element, must refer to the id attribute of the <datalist> element.
+     */
+    function form_datalist($name, $value, $options)
+    {
+        $data = [
+             'type' => 'text',
+             'name' => $name,
+             'list' => $name.'_list',
+             'value' => $value
+        ];
+        
+        $out = form_input($data)."\n";
+        $out .= "<datalist id='".$name.'_list'."'>";
+        
+        foreach ($options as $option)
+        {
+            $out .= "<option value='$option'>"."\n";
+        }
+        
+        $out .= "</datalist>"."\n";
+        
+        return $out;
+    }
+}
+
+// ------------------------------------------------------------------------
+
+if ( ! function_exists('form_keygen'))
+{
+    function form_keygen()
```

Comment:
BTW I closed one PR and opened a new one because some files from other branches "slided" into this one. I didn't know how to remove them. :-(